third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -0,0 +1,1453 @@
+//! includes specifically the following rules:
+//!
+//! (a) If immutable references to a location exist at a given program point, no mutable
+//!     references can co-exist.
+//! (b) Only one mutable reference can exist at time for the same location.
+//! (c) A location which is borrowed cannot be updated, and the value in it cannot be moved out.
+//! (d) References returned from a function call most be derived from parameters
+//!
+//! Hereby, selection of fields leads to independent 'sub-locations': the above rules are
+//! applying independently for such sub-locations. Thus one can have `&s.f` and `&mut s.g` at the
+//! same time.
+
+use crate::pipeline::livevar_analysis_processor::{LiveVarAnnotation, LiveVarInfoAtCodeOffset};
+use codespan_reporting::diagnostic::Severity;
+use im::ordmap::Entry;
+use itertools::Itertools;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{
+    ast::TempIndex,
+    model::{FieldId, FunctionEnv, Loc, QualifiedInstId, StructId},
+    ty::Type,
+};
+use move_stackless_bytecode::{
+    dataflow_analysis::{DataflowAnalysis, TransferFunctions},
+    dataflow_domains::{AbstractDomain, JoinResult, MapDomain, SetDomain},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AssignKind, AttrId, Bytecode, Operation},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::{
+    collections::{btree_map, BTreeMap, BTreeSet},
+    fmt::{Display, Formatter},
+};
+
+// ===============================================================================
+// Memory Safety Analysis
+
+// -------------------------------------------------------------------------------------------------
+// Program Analysis Domain
+
+/// Lifetime graph nodes represent information about memory locations, as well as nodes derived
+/// from that locations.
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct LifetimeNode {
+    /// Memory location associated with this node.
+    location: MemoryLocation,
+    /// Outgoing edges to descendants.
+    descendants: SetDomain<BorrowEdge>,
+    /// Backlinks to ancestors.
+    ancestors: SetDomain<LifetimeLabel>,
+}
+
+/// A label for a lifetime node.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Copy, Debug)]
+struct LifetimeLabel(usize);
+
+/// A memory location, either a global in storage or a local on the stack.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Debug)]
+enum MemoryLocation {
+    /// The associated memory is in global storage.
+    Global(QualifiedInstId<StructId>),
+    /// The associated memory is a local on the stack.
+    Local(TempIndex),
+    /// The associated memory has been reused, but the old value is still bound here to this node.
+    /// This happens after an update to the location.
+    Replaced,
+}
+
+/// Represents an edge in the lifetime graph.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Debug)]
```

Comment:
This structure involves a lot of implicit information.  You should mention that the source `LifetimeNode` of the edge is contextually defined, as the `BorrowEdge` is a "descendant" of a `LifetimeNode`.  Also, the target `LifetimeLabel` is really an index for the target `LifetimeNode` (defined in the `LifetimeState`).

Reply:
It took me several minutes to figure it out.  Code should not be a puzzle to be solved.  With a bit of commenting this could be more readable and maintainable.

Reply:
Ok, I added at the beginning the following paragraph:

```
/// The borrow graph consists of a set of `LifetimeNode` values which are labeled by
/// `LifetimeLabel`. Each node carries information about an associated
/// `MemoryLocation`, as well as the children of the node, given by list of `BorrowEdge`s.
/// The node also has backlinks to its parents, given by a set of `LifetimeLabel`, for
/// more flexible navigation through the (acyclic) graph. 
```

Notice also there is just 20 lines above this already a conceptual discussion of the logical data structure we are building here. So a desire here is also to avoid redundancy.